single_kernel_mongo/managers/cluster.py

```diff
@@ -0,0 +1,371 @@
+        try:
+            self.remove_users(relation)
+        except PyMongoError:
+            raise DeferrableError("Trouble removing router users")
+
+        if not self.charm.unit.is_leader():
+            return
+
+        logger.info("Cleaning database and user removed for mongos application")
+        self.state.secrets.remove(Scope.APP, AppPeerDataKeys.username.value)
+        self.state.secrets.remove(Scope.APP, AppPeerDataKeys.password.value)
+
+    def update_users(self):
+        """Updates users after being initialised."""
+        # VM Mongos Charm is not in charge of its users because it is a
+        # subordinate charm so we delegate everything to the MongoDB config
+        # server.
+        if self.substrate != Substrates.K8S:
+            return
+
+        # We are a Kubernetes Mongos Charm so we are in charge of our client
+        # applications and their users and we proceed to update the users and their DBs.
+        try:
+            for relation in self.state.client_relations:
+                self.dependent.mongo_manager.reconcile_mongo_users_and_dbs(relation)
+        except PyMongoError:
+            raise DeferrableError("Failed to add users on mongos-k8s router.")
+
+    def remove_users(self, relation: Relation):
+        """Handles the removal of all client mongos-k8s users and the mongos-k8s admin user.
+
+        Raises:
+            PyMongoError
+        """
+        # VM Mongos Charm is not in charge of its users because it is a
+        # subordinate charm so we delegate everything to the MongoDB config
+        # server.
+        if self.substrate != Substrates.K8S:
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        # We are a Kubernetes Mongos Charm so we are in charge of our client
+        # applications and their users and we proceed to remove the users we manage and their DBs.
+        for relation in self.state.client_relations:
+            self.dependent.mongo_manager.remove_user(relation)
+            data_interface = DatabaseProviderData(self.model, relation.name)
+            fields = data_interface.fetch_my_relation_data([relation.id])[relation.id]
+            data_interface.delete_relation_data(relation.id, list(fields.keys()))
+            secret_id = json.loads(
+                data_interface.fetch_relation_field(relation.id, "data") or "{}"
+            )["secret-user"]
+
+            user_secrets = self.charm.model.get_secret(id=secret_id)
+            user_secrets.remove_all_revisions()
+            user_secrets.get_content(refresh=True)
+            relation.data[self.charm.app].clear()
+
+    def is_ca_compatible(self) -> bool:
+        """Returns true if both the mongos and the config-server use the same CA.
+
+        Using the same CA is a requirement for sharded clusters.
+        """
+        config_server_relation = self.state.mongos_cluster_relation
+        if not config_server_relation:
```

Comment:
nit: walrus

Reply:
Even better since the variable is not use elswhere, just fully remove it